### ISSUE TITLE
Internal error when attempting to do an array comparison within square brackets

### ISSUE THREAD

**Issue by mcdobe100** [chapel-lang/chapel#15833]
### Summary of Problem
When attempting to do an array comparison within an array, there is an internal assertion error generated.

### Steps to Reproduce

**Source Code:**

```chapel
var comp = [[1,2,3] == [1,2,3]];
```

**Compile command:**
`chpl test/arrays/bugs/innerArrComp.chpl`

**Associated Future Test(s):**

[`/test/arrays/bugs/innerArrComp.future`](
           https://github.com/chapel-lang/chapel/blob/master/test/arrays/bugs/innerArrComp.future
          ) #15834  -->

### Configuration Information

- chpl version 1.23.0 pre-release (881e7d104d)
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
`CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none`
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
`Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.3.0
Thread model: posix`

**Comment by damianmoz** [comment]
The operation
```chapel
var comp = [ 1, 2, 3 ] == [ 1, 2 ];
```
will generate a run-time error with an out of bounds check because you are trying to compare a 3-element 1D array against a 2-element 1D array.

You have a second level of brackets, i.e. comp is a 1-element, 1D array of an error condition. I do not understand what you are trying to achieve.

I suggest
```chapel
proc main
{
    var t1 = [ 1, 2, 3 ];
    var t2 = [ 1, 2, 4 ];
    var comp = t1 == t2;
    var z = [ comp ];
    var y = [ t1 == t2 ];

    writeln("comp = ", comp);
    writeln("t1 = ", t1);
    writeln("t2 = ", t2);
}
```
This time, only the last assignment, that to **y**, generates the compiler error. Get rid of that line and it works.

I have not solved the problem, but I think this simpler, albeit longer, assignment might make it easier to track the compiler error. My 2c.

**Comment by ty1027** [comment]
In any code (regardless of whether it is written nicely or poorly), isn't an "internal error" a bug of the compiler to be reported (and so hopefully to be fixed)...?

By the way, this line also results in the same internal error:
```
var comp = [ [1,2,3] == [1,2,4] ];
// var comp = [ ( [1,2,3] == [1,2,4] ) ];  // parens have no effect here...

$CHPL_HOME/modules/internal/ChapelBase.chpl:983:
internal error: RES-FUN-ION-9910 chpl version 1.22.0[1,2]
Note: This source location is a guess.
```

while this code gives this result:

```
var comp = [ [true, true, false] ];
writeln( "comp = ", comp );  // true true false

writeln( "comp.type = ", comp.type:string );
// [domain(1,int(64),false)] [domain(1,int(64),false)] bool

writeln( [1,2,3] == [1,2,4] );  // true true false
```

Further, this code
```
writeln( [1,2,3,4,5] == [1,2] );
```
gives a runtime error,
```
error: zippered iterations have non-equal lengths
```

while a similar code (with a temporary variable)
```
var comp = [1,2,3,4,5] == [1,2];
// var comp = ( [1,2,3,4,5] == [1,2] );  // same result
writeln( comp );
```
compiles but gives an unexpected (?) result (`true true false false false`)...

**Comment by damianmoz** [comment]
I was not inferring that it was **not** a compiler error, not that this error does not need to be fixed. I was trying to suggest providing several variations of the most compact program that causes the error. This might help the compiler team track down the problem.

Byt the way, with Chapel 1.20, your last chunk of code that use a temporary variable and a writeln,, compiles but gives the more correct error message:
```
3: error: halt reached - array index out of bounds: (5)
```
Again, this (correctly) s a run-time error.

The problem is when it is the compiler doing the evaluation, trapping the error, and identifying the problem. That is a lot more work.

**Comment by bradcray** [comment]
> isn't an "internal error" a bug of the compiler to be reported (and so hopefully to be fixed)...?

Definitely.

> By the way, with Chapel 1.20, your last chunk of code that use a temporary variable and a writeln,, compiles but gives the more correct error message:

Well that's concerning... that's a backslide that doesn't feel familiar to me.  I'm looking into when this crept in.  Thanks for mentioning in @damianmoz.

**Comment by bradcray** [comment]
This change for the worse in behavior unfortunately came in as part of PR #13767 which had other virtues.  :'(  It was noted in issue #14264, but unfortunately escaped my notice until now.

**Comment by bradcray** [comment]
This issue was resolved by PR #16802.  I'm retiring the future in PR #17409.
